selfservice/strategy/saml/handler.go

```diff
@@ -0,0 +1,374 @@
+package saml
+
+import (
+	"bytes"
+	"context"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"encoding/xml"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	dsig "github.com/russellhaering/goxmldsig"
+
+	"github.com/crewjam/saml/samlsp"
+	"github.com/julienschmidt/httprouter"
+
+	"github.com/ory/herodot"
+	"github.com/ory/kratos/continuity"
+	"github.com/ory/kratos/driver/config"
+	"github.com/ory/kratos/selfservice/errorx"
+
+	samlidp "github.com/crewjam/saml"
+
+	"github.com/ory/kratos/session"
+	"github.com/ory/kratos/x"
+	"github.com/ory/x/decoderx"
+	"github.com/ory/x/fetcher"
+	"github.com/ory/x/jsonx"
+)
+
+var ErrNoSession = errors.New("saml: session not present")
+
+var samlMiddlewares = make(map[string]*samlsp.Middleware)
+
+type ory_kratos_continuity struct{}
+
+type (
+	handlerDependencies interface {
+		x.WriterProvider
+		x.CSRFProvider
+		session.ManagementProvider
+		session.PersistenceProvider
+		errorx.ManagementProvider
+		config.Provider
+	}
+	HandlerProvider interface {
+		LogoutHandler() *Handler
+	}
+	Handler struct {
+		d  handlerDependencies
+		dx *decoderx.HTTP
+	}
+)
+
+type SessionData struct {
+	SessionID string
+}
+
+func NewHandler(d handlerDependencies) *Handler {
+	return &Handler{
+		d:  d,
+		dx: decoderx.NewHTTP(),
+	}
+}
+
+func (h *Handler) RegisterPublicRoutes(router *x.RouterPublic) {
+	h.d.CSRFHandler().IgnoreGlob(RouteBaseAcs + "/*")
+
+	router.GET(RouteMetadata, h.serveMetadata)
+	router.GET(RouteAuth, h.loginWithIdp)
+}
+
+// Handle /selfservice/methods/saml/metadata
+func (h *Handler) serveMetadata(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	config := h.d.Config()
+	pid := ps.ByName("provider")
+
+	if samlMiddlewares[pid] == nil {
+		if err := h.instantiateMiddleware(r.Context(), *config, pid); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
```

Comment:
This is going to be a 200 JSON response or a 303 redirect, but don't we expect XML here?